src/main/kotlin/org/rust/ide/actions/macroExpansion/macroExpansionViewUtils.kt

```diff
@@ -0,0 +1,130 @@
+
+    fun expandAndShowMacro(project: Project, editor: Editor, macroToExpand: RsMacroCall) {
+        val expansions = expandMacroCall(macroToExpand) ?: return
+
+        if (expansions.isEmpty()) return
+
+        val component = MacroExpansionViewComponent(expansions)
+
+        val createPopup = JBPopupFactory.getInstance().createComponentPopupBuilder(component, component)
+            .setProject(project)
+            .setResizable(true)
+            .setMovable(true)
+            .setRequestFocus(true)
+            .setTitle(getMacroExpansionPopupTitle(macroToExpand))
+            .createPopup()
+
+        PopupPositionManager.positionPopupInBestPosition(createPopup, editor, null)
+    }
+
+    private fun expandMacroCall(macroToExpand: RsMacroCall): List<RsExpandedElement>? {
+        if (!expandRecursively) return macroToExpand.expansion
+
+        val expansionText = macroToExpand.expandAllMacrosRecursively() ?: return null
+        val expansionElements = RsPsiFactory(macroToExpand.project)
+            .parseExpandedTextWithContext(macroToExpand, expansionText)
+
+        return expansionElements
+    }
+
+    private fun getMacroExpansionPopupTitle(macroToExpand: RsMacroCall): String {
+        return if (expandRecursively) {
+            "Recursive expansion of ${macroToExpand.text} macro"
+        } else {
+            "First level expansion of ${macroToExpand.text} macro"
+        }
+    }
+
+}
+
+/**
+ * Simple view to show some code. Inspired by [com.intellij.codeInsight.hint.ImplementationViewComponent].
+ */
+private class MacroExpansionViewComponent(expansions: List<RsExpandedElement>) : JPanel(BorderLayout()) {
+
+    init {
+        require(expansions.isNotEmpty()) { "Must be at least one expansion!" }
+
+        val firstExpansion = expansions.first()
+        val project = firstExpansion.project
+
+        val reformatted = expansions.map { project.formatPsiElement(it) }
+        val editor = project.createReadOnlyEditorWithElements(reformatted)
+
+        setupSimpleEditorLook(editor)
+        add(editor.component)
+
+        editor.highlighter = project.getHighlighterForPsiElement(firstExpansion)
+    }
+
+    private fun setupSimpleEditorLook(editor: EditorEx) {
+        with(editor.settings) {
+            additionalLinesCount = 1
+            additionalColumnsCount = 1
+            isLineMarkerAreaShown = false
+            isIndentGuidesShown = false
+            isLineNumbersShown = false
+            isFoldingOutlineShown = false
+        }
+    }
+
+}
+
+private fun Project.formatPsiElement(element: PsiElement): PsiElement =
+    CodeStyleManager.getInstance(this).reformatRange(
+        element,
+        element.textRange.startOffset,
+        element.textRange.endOffset
+    )
+
+private fun Project.createReadOnlyEditorWithElements(expansions: List<PsiElement>): EditorEx {
+    val factory = EditorFactory.getInstance()
+
+    val text = expansions.joinToString("\n") { it.text }
```

Comment:
Does `\n` works on Windows? If not, probably you should use `System.getProperty("line.separator")`.